server/src/controllers/ChapterUser/resolver.ts

```diff
@@ -14,9 +14,71 @@ import { ResolverCtx } from '../../common-types/gql';
 import { prisma } from '../../prisma';
 import { ChapterUser, UserBan } from '../../graphql-types';
 import { Permission } from '../../../../common/permissions';
+import { ChapterRoles } from '../../../prisma/generator/factories/chapterRoles.factory';
+import { InstanceRoles } from '../../../prisma/generator/factories/instanceRoles.factory';
 
 const UNIQUE_CONSTRAINT_FAILED_CODE = 'P2002';
 
+const chapterUsersInclude = {
+  chapter_role: {
+    include: {
+      chapter_role_permissions: { include: { chapter_permission: true } },
+    },
+  },
+  user: true,
+};
+
+type UserWithUserChapters = Prisma.chapter_usersGetPayload<{
+  include: {
+    chapter_role: {
+      include: {
+        chapter_role_permissions: { include: { chapter_permission: true } };
+      };
+    };
+    user: {
+      include: {
+        instance_role: true;
+        user_chapters: { include: { chapter_role: true } };
+      };
+    };
+  };
+}>;
+
+interface ChangeInstanceRoleData {
+  chapterUser: UserWithUserChapters;
+  newChapterRole: string;
+  oldInstanceRole: string;
+}
+
+const getInstanceRoleName = ({
+  chapterUser,
+  newChapterRole,
+  oldInstanceRole,
+}: ChangeInstanceRoleData) => {
+  if (oldInstanceRole === InstanceRoles.owner) return oldInstanceRole;
+
+  if (
+    newChapterRole === ChapterRoles.administrator &&
+    oldInstanceRole !== InstanceRoles.chapter_administrator
+  ) {
+    return InstanceRoles.chapter_administrator;
+  }
+
+  if (
+    newChapterRole === ChapterRoles.member &&
+    oldInstanceRole === InstanceRoles.chapter_administrator
+  ) {
+    const isStillAdmin = chapterUser.user.user_chapters.some(
+      (chapter_user) =>
+        chapter_user.chapter_id !== chapterUser.chapter_id &&
+        chapter_user.chapter_role.name === ChapterRoles.administrator,
+    );
+
+    if (!isStillAdmin) return InstanceRoles.member;
+  }
+  return oldInstanceRole;
+};
```

Comment:
I think this deserves one or two unit tests.  It looks solid, but there's enough complexity that I think some simple checks would help.  Plus it's a nice, well-behaved function with no hidden dependencies, so it's easy to test!

Reply:
I've most likely went a bit overboard with the unit tests...

I had issues with making appropriate `chapterUser` objects for tests, so I've refactored it to use simpler type as argument. Names of both these functions are a bit iffy, but so far I haven't found anything short for it.

Reply:
> I've most likely went a bit overboard with the unit tests...

Not necessarily.  I generally aim for 100% coverage when I'm creating a new function, just so I know when to stop.  If what you've added is necessary to get to 100% that's totally fine. 

If you can cut a few tests/fixtures and still be at 100%, that's helpful (since there's less overall to maintain), but it's not too crucial.

> I had issues with making appropriate chapterUser objects for tests, so I've refactored it to use simpler type as argument.

As long as there's no need to assert types, it's fine.